verifier/src/evidence.rs

```diff
@@ -0,0 +1,334 @@
+/// * `accessor_method` - The method on `ReportBody` that returns the field
+macro_rules! quote_application_report_body_field_accessor {
+    ($($field_type:ty, $accessor_method:ident;)*) => {$(
+        impl<'a, Q: AsRef<[u8]>> Accessor<$field_type> for Evidence<'a, Q> {
+            fn get(&self) -> $field_type {
+                self.quote.app_report_body().$accessor_method()
+            }
+        }
+    )*}
+}
+
+quote_application_report_body_field_accessor! {
+    Attributes, attributes;
+    ConfigId, config_id;
+    ConfigSvn, config_svn;
+    CpuSvn, cpu_svn;
+    ExtendedProductId, isv_extended_product_id;
+    FamilyId, isv_family_id;
+    IsvProductId, isv_product_id;
+    IsvSvn, isv_svn;
+    MiscellaneousSelect, miscellaneous_select;
+    MrEnclave, mr_enclave;
+    MrSigner, mr_signer;
+    ReportData, report_data;
+}
+
+#[cfg(test)]
+mod test {
+    extern crate alloc;
+    use super::*;
+    use crate::{
+        AdvisoriesVerifier, AdvisoryStatus, And, AttributesVerifier, MrSignerVerifier,
+        Quote3Verifier, TcbInfoRawVerifier, VerificationTreeDisplay, Verifier,
+    };
+    use alloc::format;
+    use core::mem;
+    use der::DateTime;
+    use mc_sgx_dcap_sys_types::{sgx_ql_ecdsa_sig_data_t, sgx_quote3_t};
+    use p256::ecdsa::VerifyingKey;
+
+    fn tcb_signing_key() -> VerifyingKey {
+        let pem = include_str!("../data/tests/tcb_signer.pem");
+        let certificate = Certificate::from_pem(pem).expect("failed to parse PEM");
+        let key = VerifyingKey::from_sec1_bytes(
+            certificate
+                .tbs_certificate
+                .subject_public_key_info
+                .subject_public_key
+                .as_bytes()
+                .expect("Failed to parse public key"),
+        )
+        .expect("Failed to decode public key");
+
+        key
+    }
+
+    fn quote_signing_key<T: AsRef<[u8]>>(quote: &Quote3<T>) -> VerifyingKey {
+        let signature_data = quote.signature_data();
+        let certification_data = signature_data.certification_data();
+        let CertificationData::PckCertificateChain(chain) = certification_data else {
+            panic!("Unexpected certification data type");
+        };
+        let leaf_pem = chain.into_iter().next().expect("No leaf certificate");
+        let certificate =
+            Certificate::from_pem(leaf_pem).expect("Failed to parse leaf certificate");
+
+        let key = VerifyingKey::from_sec1_bytes(
+            certificate
+                .tbs_certificate
+                .subject_public_key_info
+                .subject_public_key
+                .as_bytes()
+                .expect("Failed to parse public key"),
+        )
+        .expect("Failed to decode public key");
+
+        key
+    }
+
+    type VerifierType<T> = And<
+        Quote3Verifier<T>,
+        And<MrSignerVerifier, And<AttributesVerifier, And<TcbInfoRawVerifier, AdvisoriesVerifier>>>,
+    >;
```

Comment:
I'm not fond of this. It may be my lack of experience, but I wasn't able to get the compiler to infer the type for the below function, so punted and chose to be more explicit.